zingolib/src/wallet/notes/transparent.rs

```diff
@@ -0,0 +1,167 @@
+        } else {
+            (None, None)
+        };
+
+        zcash_encoding::Vector::write(&mut writer, &self.script, |w, b| w.write_all(&[*b]))?;
+
+        zcash_encoding::Optional::write(&mut writer, spent, |w, transaction_id| {
+            w.write_all(transaction_id.as_ref())
+        })?;
+
+        zcash_encoding::Optional::write(&mut writer, spent_at_height, |w, s| {
+            w.write_i32::<byteorder::LittleEndian>(s)
+        })?;
+
+        Ok(())
+    }
+    pub fn read<R: std::io::Read>(mut reader: R) -> std::io::Result<Self> {
+        let version = reader.read_u64::<byteorder::LittleEndian>()?;
+
+        let address_len = reader.read_i32::<byteorder::LittleEndian>()?;
+        let mut address_bytes = vec![0; address_len as usize];
+        reader.read_exact(&mut address_bytes)?;
+        let address = String::from_utf8(address_bytes).unwrap();
+        assert_eq!(address.chars().take(1).collect::<Vec<char>>()[0], 't');
+
+        let mut transaction_id_bytes = [0; 32];
+        reader.read_exact(&mut transaction_id_bytes)?;
+        let transaction_id = TxId::from_bytes(transaction_id_bytes);
+
+        let output_index = reader.read_u64::<byteorder::LittleEndian>()?;
+        let value = reader.read_u64::<byteorder::LittleEndian>()?;
+        let _height = reader.read_i32::<byteorder::LittleEndian>()?;
+
+        let script = zcash_encoding::Vector::read(&mut reader, |r| {
+            let mut byte = [0; 1];
+            r.read_exact(&mut byte)?;
+            Ok(byte[0])
+        })?;
+
+        let spent = zcash_encoding::Optional::read(&mut reader, |r| {
+            let mut transaction_bytes = [0u8; 32];
+            r.read_exact(&mut transaction_bytes)?;
+            Ok(TxId::from_bytes(transaction_bytes))
+        })?;
+
+        let spent_at_height = if version <= 1 {
+            None
+        } else {
+            zcash_encoding::Optional::read(&mut reader, |r| {
+                r.read_i32::<byteorder::LittleEndian>()
+            })?
+        };
+
+        let _unconfirmed_spent = if version == 3 {
+            zcash_encoding::Optional::read(&mut reader, |r| {
+                let mut transaction_bytes = [0u8; 32];
+                r.read_exact(&mut transaction_bytes)?;
+
+                let height = r.read_u32::<byteorder::LittleEndian>()?;
+                Ok((TxId::from_bytes(transaction_bytes), height))
+            })?
+        } else {
+            None
+        };
+
+        let spent_tuple: Option<(TxId, u32)> = if let Some(txid) = spent {
+            if let Some(height) = spent_at_height {
+                Some((txid, height as u32))
+            } else {
+                Some((txid, 0))
+            }
+        } else {
+            None
+        };
+
+        Ok(TransparentNote {
+            address,
+            txid: transaction_id,
+            output_index,
+            script,
+            value,
+            spent: spent_tuple,
+            unconfirmed_spent: None,
+        })
+    }
+}
```

Comment:
not sure why read and write were swapped in order but not an issue, just harder to review and dont see the benefit

Reply:
If this was the result of an automation we should configure the tool to enforce the preferred order.

Reply:
sorry for any confusion. i feel strongly that write comes before read in a flow